File: LeetCode-December-Challenge/PossibleBipartition.java
import java.util.*;
class Solution {
    public boolean possibleBipartition(int n, int[][] dislikes) {
        ArrayList<Integer>[] adj = new ArrayList[n + 1];
        for(int i = 0; i <= n; ++i)
            adj[i] = new ArrayList<>();
        int[] color = new int[n + 1];
        boolean[] visited = new boolean[n + 1];
        for(int i = 0; i < dislikes.length; ++i){
            int u = dislikes[i][0];
            int v = dislikes[i][1];
            adj[u].add(v);
            adj[v].add(u);
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i <= n; ++i){
            if(!visited[i]){
                color[i] = 1;
                q.add(i);
                while(!q.isEmpty()){
                    int curr = q.poll();
                    if(visited[curr]) continue;
                    visited[curr] = true;
                    for(int x: adj[curr]){
                        if(color[x] == color[curr]) return false;
                        else if(color[curr] == 1) color[x] = 2;
                        else color[x] = 1;
                        q.add(x);
                    }
                }
            }
        }
        return true;
    }
}
public class PossibleBipartition {
    public static void main(String[] args) {
        int n = 4;
        int[][] dislikes = {{1,2},{1,3},{2,4}};
        Solution s = new Solution();
        System.out.println(s.possibleBipartition(n, dislikes));
    }
}

File: Year-2022/LeetCode-December-Challenge/PossibleBipartition.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        vector<vector<int>> adj(n + 1);
        vector<int> color(n + 1, 0);
        vector<bool> visited(n + 1, false);
        for(auto &x: dislikes){
            int u = x[0];
            int v = x[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        queue<int> q;
        for(int i = 0; i <= n; ++i){
            if(!visited[i]){
                color[i] = 1;
                q.push(i);
                while(!q.empty()){
                    int curr = q.front();
                    q.pop();
                    if(visited[curr]) continue;
                    visited[curr] = true;
                    for(auto x: adj[curr]){
                        if(color[x] == color[curr]) return false;
                        else if(color[curr] == 1) color[x] = 2;
                        else color[x] = 1;
                        q.push(x);
                    }
                }
            }
        }
        return true;
    }
};
int main(){
    int n = 4;
    vector<vector<int>> dislikes = {{1,2},{1,3},{2,4}};
    Solution *s = new Solution();
    cout<<boolalpha<<s->possibleBipartition(n, dislikes)<<endl;
    return 0;
}